### DISCUSSION TITLE
Symlink inside Folder public directory

### DISCUSSION THREAD

**Discussion by nguyenbatranvan** [nguyenbatranvan/vite-multiple-assets#28] [Ideas]
## Currently

In the shared folder, if there is a symbolic link to a folder or file, in the development environment, the library will not be able to read that link and when building, the symlink will also be removed.

## Solution

Will add an option for users to pass in symbolic links. This will read and find the targetPath of the symlink. Read and replace it in development mode and write its absolute path to the build output.

**Comment by nguyenbatranvan** [comment]
@FarhanMS123 Are you interested in this new feature. I have an idea for this. Users will add an option called `followSymlinks` that will look like this:
```
// type symlinks
type TSymlink = {
    symlink: string;
}

```

`vite.config.ts`

```
// type followSymlinks:TSymlink[];

DynamicPublicDirectory([
            {
                input: "../../../{\x01,shared-assets}/**",
                output: "/shared/images",
                watch: true // default
            }, "public/**", "{\x01,public2}/**"], {
            followSymlinks: [{
                symlink: '../../../shared-assets/change.png' // symlink of a/b/c/d/e/f/change.png
            }]
        })
```

`symlink` is a symbolic link that the user must pass if using the `followSymlinks` option

**Reply by FarhanMS123** [reply]
This seems a bit complex for me. When I try to integrate `fast-glob`, I'm trying to met with your code style, simple options simple use. I suggest to implement `followSymlinks` flag like how `fast-glob` do. What do you think?

**Reply by nguyenbatranvan** [reply]
I'm not sure if `fast-glob` works well with symlinks?

**Reply by FarhanMS123** [reply]
I try to find the issues related to `symbolic` and `symlink`, and all those already closed except this one: https://github.com/mrmlnc/fast-glob/issues/300 and I think this not relate with out problem.

I just think we could trust `fast-glob` by whatever logic it has behind. and give it a chance. We can use `followSymbolicLinks` and enforce `throwErrorOnBrokenSymbolicLink` to `true` and then we can handle the error by our logic.

- https://github.com/mrmlnc/fast-glob?tab=readme-ov-file#followsymboliclinks

**Reply by nguyenbatranvan** [reply]
I tried with `fast-glob`, it only reads `symlinks` with the options below. I want to emphasize that `onlyFiles = false`, different from before:
```
const files = await fg.glob(__transformFiles, {
        onlyFiles: false,
        onlyDirectories: false,
        throwErrorOnBrokenSymbolicLink:true,
        markDirectories: false,
        followSymbolicLinks:false,
    });
   ```

**Reply by nguyenbatranvan** [reply]
not stopping there there are 2 problems I see if applying fast glob:
- You will not be able to use `symlinks` directly in the image, instead, you have to rewrite the image path via `ServerMiddleWare`
- Cannot create folders using `fs.mkdir`, requires some editing. As described below

```
fs.mkdir(dirPath, { recursive: true }, (err) => {
  if (err) {
    return console.error('Error creating directory:', err);
  }
  console.log('Directory created successfully!');

  // Create the symbolic link to the directory
  fs.symlink(dirPath, symlinkPath, 'dir', (err) => {
    if (err) {
      return console.error('Error creating symlink:', err);
    }
    console.log('Symlink to directory created successfully!');
  });
});
```

**Reply by nguyenbatranvan** [reply]
i have create [PR](https://github.com/nguyenbatranvan/vite-multiple-assets/pull/31/files) for symlinks. Currently it only works with build & production environments. Thanks for your discussion, I think it's simpler to just add `followSymlinks` (true | false) at [here](https://github.com/nguyenbatranvan/vite-multiple-assets/pull/31/files#diff-73c3ec5f8c4aba962fcf016e5d46dc983eec91f6ef23c951970896428275f2a7L76)!!

**Reply by nguyenbatranvan** [reply]
i'm done!! U can check

**Reply by FarhanMS123** [reply]
Ok. Give me some times to check.

**Reply by FarhanMS123** [reply]
Hi! Sorry, seems I have no time for this. I may do major improvement (and maybe refactoring) if there is requirements I want to apply on my side. But for now, I already use this plugin to few of my project, and seems there is no conflicted to those base on my original use-case. So, I think it seems good.

**Comment by FarhanMS123** [comment]
Hi! I actually put that info in here: https://nguyenbatranvan.github.io/vite-multiple-assets-doc/guide/index.html#enforced-behaviors

All `opts` would be passed to `fast-glob` directly. So, without changing the code, it would works. Of course with a cost that their linter would scream out.

For us, we only need to add `followSymlinks` in this Pick type: https://github.com/nguyenbatranvan/vite-multiple-assets/blob/main/packages/libs/src/types.d.ts#L76
